content/blog/design-kgateway-for-scalability.md

```diff
@@ -0,0 +1,170 @@
+author: Lin Sun & Yuval Kohavi
+---
+
+With the Kubernetes Gateway API becoming the de facto standard for managing traffic into, out of, and within clusters, a growing number of gateways now implement this API. While gateways are often expected to "just work," choosing the right one can have major implications—especially at scale.
+
+Before comparing features, the most critical consideration is whether a gateway is built on a solid, reliable foundation–the kind of foundation that just works.
+
+## Aren’t All Envoy-Based Gateways the Same?
+
+It's a fair question. Many gateways use the Envoy proxy under the hood to enforce routing, security, and other traffic policies. But despite sharing the same data plane, not all gateways are equal—because the **control plane** makes all the difference.
+
+The control plane is responsible for translating Kubernetes Gateway API resources into actual Envoy configuration. This translation layer can be simple for a few routes, but when you’re managing 20,000 resources (which will translate into 500,000+ lines of Envoy config), the efficiency and scalability of the control plane become critical—as I noted in a previous [LinkedIn post](https://www.linkedin.com/posts/lin-sun-a9b7a81_from-20000-kubernetes-gateway-resources-activity-7305695363049373696-UVNz?utm_source=share&utm_medium=member_desktop&rcm=ACoAAABLihcBuozqLyftNtauegAdN2-QszsmqQQ).
+
+## Designing kgateway for Scalability
+
+When we started building **Gloo** (now **kgateway**) seven years ago, we used a snapshot-based model that recalculated everything on every update—whether it was a new route, an update, or a backend change. This meant even small changes triggered a full control plane recalculation. With Kubernetes, where pods and backends change constantly, this was not scalable.
+
+Managing resource dependencies via manual references like `targetRefs` and `extensionRefs` also proved cumbersome. Dependency resolution sometimes failed, creating reliability issues.
+
+Learning from that experience, we designed **kgateway** using the battle-tested [krt](https://github.com/istio/istio/blob/master/pkg/kube/krt/README.md) framework to handle dependency tracking automatically. Now, only the **affected objects** are re-translated when a change occurs—ensuring fast updates and efficient scaling.
+
+Let’s walk through a few core scenarios to illustrate how kgateway scales to support large teams and applications—starting with a setup of 10,000 routes and backends—and see how kgateway handles new routes, route update, deletion or control plane becomes unavailable.
+
+## Core Scalability Scenarios
+
+### 1. When a New Route Is Added
+
+In a production system with thousands of existing routes, adding a new one should be near-instant. Ideally, the route becomes effective within 1 second, without needing a periodic timer or restart. This ensures teams can test and deploy with confidence.
+
+{{< reuse-image src="blog/scalability-add-a-route.gif" width="750px" >}}
```

Comment:
You can't see what's on screen in these GIFs, and they cycle very quickly.   It was only when I got to the end I realised that the video was included too.  Want to call that out?